031-tls-encrypt-the-kafka-connect-rest-api.md

```diff
@@ -0,0 +1,168 @@
+
+For this proposal, the `protocol` field only supports two values: `http` and `https`. If the `restListeners` list is empty the `KafkaConnect` and `KafkaMirrorMaker2` 
+runtimes will be created with a single unencrypted listener on 8083 to match the existing behaviour.
+
+The `restCertChainAndKey` field will be required if the `https` protocol is selected.
+
+For example, the following CR will enable the HTTP listener on port 8083 and the HTTPS listener on port 8443:
+```
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+spec:
+  restListeners:
+  - protocol: http
+  - protocol: https
+  ...
+```
+
+The schema for `restListeners` will be as follows:
+```
+openAPIV3Schema:
+  type: object
+  properties:
+    spec:
+      type: object
+      properties:
+        restListeners:
+          type: array
+          items:
+            type: object
+            properties:
+              protocol:
+                type: string
+                enum:
+                - http
+                - https
+                description: The protocol of the REST listener.
+              restCertChainAndKey:
+                type: object
+                properties:
+                  certificate:
+                    type: string
+                    description: The name of the ca certificate file in the Secret.
+                  key:
+                    type: string
+                    description: The name of the ca private key file in the Secret.
+                  secretName:
+                    type: string
+                    description: The name of the Secret containing the certificate.
+                  required:
+                    - key
+                    - certificate
+                    - secretName
+                description: CA certificate to use to generate.
+            required:
+            - protocol
+          description: List of additional REST listeners.
+        ...
+```
+
+the above CR would result in the following properties in the generated connect configuration:
+
+```
+listeners: https://:8443,http://:8083
+rest.advertised.listener: https
+rest.advertised.port: 8443
+listeners.https.ssl.client.auth: none
+listeners.https.ssl.truststore.location: /tmp/kafka/kafka-connect-rest.truststore.p12
+listeners.https.ssl.truststore.password: ***generated password***
+listeners.https.ssl.truststore.type: PKCS12
+listeners.https.ssl.keystore.location: /tmp/kafka/kafka-connect-rest.keystore.p12
+listeners.https.ssl.keystore.password: ***generated password***
+listeners.https.ssl.keystore.type: PKCS12
+```
+
+### Certificates
+
+If the user decides to enable an encrypted HTTPS listener, they must have the `restCertChainAndKey` 
+property present in the `KafkaConnect` CR. The property must point to an existing secret that contains the 
+CA certificate and CA private key that they want Kafka Connect to use. This secret can be created by the 
+user, or it can be the [Strimzi cluster CA secret](#using-the-strimzi-cluster-ca).
```

Comment:
I realise this is tricky, but do we _really_ want Strimzi to get deeper into the business of issuing its own certificates like this? It's often not what people operating Strimzi want: They often want to use things like vault or public CAs to issue a server certificate, in which case there is no CA certificate that Strimzi can access to issue itself a server certificate.

An alternative model would be just to point to a server certificate and key in the `KafkaConnect` or MM CRs, and assume that this is issued by some unspecified means outside the system. The CO could take care to restart the Pod when the certificate generation changed. Then in the CO configuration (i.e. it's ENV) we could have the option of using the cluster CA to issue these things automatically. 

That would allow people to issue their own endpoint certificates however they chose, but it also wouldn't build in the assumption that Strimzi is doing the issuing. It's especially important to be careful about building these assumptions into the CRDs because those are APIs which are really difficult to evolve. Doing it in the CO config makes it a bit easier to deprecate configs at a later date, if needed.